### PR TITLE
feat(Date Components): Accept date-fns locale object

### PR DESCRIPTION
**Pull request by mdodgelooker**
## Deprecated
- `Locales` enum
- `dateFnLocaleMap` map
- `dateStringLocale` prop
- `localization` prop

## Added
- `locale` prop, accepts `date-fns` type `Locale` – replaces `dateStringLocale` and `localization` props
- `dateStringFormat` prop, supports all `date-fns` format patterns – replaces a certain use of `dateStringLocale` (i.e. how we currently use `dateStringFormat={Locales.Japanese}` in filter-components just to get the YYYY-MM-DD format
- `firstDayOfWeek` prop, in case it needs to be different from what the `date-fns` locale has set, (filter-components does)

## Next steps

1. With **v3.0** we can remove deprecated props & exports. Removing `dateFnLocaleMap` specifically will improve bundle size since we will no longer be importing any of the locales from `date-fns` besides `en`.
2. Update `@looker/filter-components` with the new props above
3. Both filter packages: stop exporting every locale's translations (except `en`) and instead support deep importing / dynamic importing those resources.
4. Explore ways for packages (`components-date`, `filter-components`, etc) to store their specific locale resources in the `ComponentsProvider` context.

### CODE REVIEW

**Comment by mdodgelooker** [comment]
Tagging @ElliotTheHuman too because you're also working on a date thing at the moment. 🙂

File: packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx [hunk 1]
```diff
@@ -26,60 +26,45 @@
...
-type DateFormats = 'short' | 'medium' | 'long' | 'full'
-
-export interface DateTimeFormatProps {
+export type DateTimeFormatProps = {
   children?: Date
-  format?: DateFormats
-  locale?: LocaleCodes
+  format?: DateFormats | string
+  // TODO delete Locales when enum is removed
+  locale?: Locale | Locales
```
Comment (lukelooker) [inline packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx:39]: Should this get marked as `@deprecated` ?

File: packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx [hunk 2]
```diff
@@ -26,60 +26,45 @@
...
-export interface ExtendedDateTimeFormatOptions
-  extends Intl.DateTimeFormatOptions {
-  dateStyle?: DateFormats
-  timeStyle?: DateFormats
+// TODO delete when Locales enum is removed
+const localeIsString = (
+  locale: DateTimeFormatProps['locale']
+): locale is Locales => {
+  return typeof locale === 'string'
 }
```
Comment (lukelooker) [inline packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx:50]: Mark as `@deprecated`?

**Review by lukelooker — APPROVED** [review]
Awesome update! The scheme makes a lot more sense after this refactor :)

File: packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx (hunk 1 above)
Comment (mdodgelooker) [reply packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx:39]: The prop will remain, it'll just get a narrower type after 3.0. Is there a way to mark that? Or should I create a new prop "localeConfig" so I can mark the old one as deprecated?

Comment (lukelooker) [reply packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx:39]: Oh. Sorry, I misread it – I don't think there's a good way to mark it as "partially deprecated" 

I think it's fine as-is for now.

Comment (mdodgelooker) [reply packages/components-date/src/DateTimeFormat/DateTimeFormat.tsx]: I just added a note in the JSDoc, if anyone reads those. 🙂 Rebased and needing another ✅

**Review by lukelooker — APPROVED** [review]

**Comment by lukelooker** [comment]
@mdodgelooker You got bitten by my prettier update. Need to rebase and update. Sorry!

**Review by lukelooker — APPROVED** [review]

**Pull request merged**